File: elixir/nth-prime/nth_prime.exs
defmodule Prime do

  @doc """
  Generates the nth prime.
  """
  @spec nth(non_neg_integer) :: non_neg_integer
  def nth(count) when count < 1, do: raise ArgumentError
  def nth(1), do: 2
  def nth(count) do
    Stream.unfold(1, &({&1, &1+2}))
    |> Stream.filter(&prime?/1)
    |> Stream.drop(count-1)
    |> Enum.take(1)
    |> Enum.at(0)
  end

  defp prime?(number) do
    all_factors(number) == [1]
  end

  defp all_factors(number) do
    all_factors(number, round(:math.sqrt(number)))
  end
  defp all_factors(number, 1), do: [1]
  defp all_factors(number, divisor) when rem(number, divisor) == 0 do
    [divisor|all_factors(number, divisor-1)]
  end
  defp all_factors(number, divisor), do: all_factors(number, divisor-1)
end

File: ruby/nth-prime/nth_prime.rb
module Prime
  module_function

  def nth(number)
    raise ArgumentError, "Number must be greater than zero" if number < 1
    return 2 if number < 2

    current = 1
    candidate = 2
    while current < number
      candidate += 1
      current += 1 if prime?(candidate)
    end
    candidate
  end

  def prime?(number)
    !!2.upto(Math.sqrt(number).floor).each do |i|
      return false if number % i == 0
    end
  end
end

module BookKeeping
  VERSION = 1
end
